### ISSUE TITLE
Release cycle and tagged versions

### ISSUE THREAD

**Issue by finnlewis** [localgovdrupal/localgov#139]
Now that councils are starting to use LocalGov Drupal, we need to establish how to tag releases for all projects and control the upgrade process. 

I guess the first step is to confirm the nomenclature for versioning. 

Do we start at 

0.0.1

Then move up from there until we have a 

1.0.0 

release?

Edit (19/04/2021)

Preparing this for sprint 11 (first sprint of beta).
We'd like to bring this into the sprint, so defining acceptance tests. 

## Acceptance tests. 

1. Agree the release tagging policy with the technical group.
2. Publish and share the policy with the wider community.
3. Make necessary changes to the composer.json files.
4. Test this against Drupal9 or dependent modules ?

## Estimates: 
 - 8

**Comment by finnlewis** [comment]
Useful: https://www.drupal.org/node/1015226

**Comment by finnlewis** [comment]
@ekes @tanc we need to think about this ASAP from the Lambeth perspective

**Comment by finnlewis** [comment]
@stephen-cox @Adnan-cds @andybroomfield any initial thoughts to add?

**Comment by Adnan-cds** [comment]
Hi Finn,
Thanks for linking to the Drupal documentation. It looks we need to decide two things:
- Should we stick to the master branch or should we create a *1.x* branch and treat that as the stable branch.  I prefer the master branch although if we decide to make a backward incompatible changes later, the 1.x could be useful.
- Tag names: these follow semantic versioning now-a-days.  That leaves the alpha/beta/rc prefixes.  Should we start using these immediately?

@finnlewis

**Comment by stephen-cox** [comment]
We should certainly user semantic versioning, so that tags follow the MAJOR.MINOR.PATCH convention. The actual versions could then be incremented as:

- MAJOR => Full release with potentially breaking changes
- MINOR => New features with non-breaking changes
- PATCH => Bug fix

If we we're to follow something like this then we need to decide who and how version upgrades happen and to what extent the versions across the different modules are kept in sync.

We could we decide something like: the product group decide on major releases, a group of developers working and reviewing on a feature can do a minor version upgrade and any two developers (fixing and reviewing a bug) can do a patch release? In this scenario only the major release versions are kept in sync for the whole distro, while each module will have it's own minor and patch release versions. I'm sure there are other things to consider and ways to do this.

I agree that we want to move to some sort of versioned releases ASAP. Maybe we start with 0.1.0 releases for all or most modules to acknowledge that they are working, but still in development. Maybe one aim of a beta phase would be a 1.0.0 release?

**Comment by ekes** [comment]
I think {major}.{minor}.{patch} is pretty much a given as it's the general and now Drupal standard. As @stephen-cox asks what these mean is most important. How we branch and tag is also a useful question.

> If we we're to follow something like this then we need to decide who and how version upgrades happen and to what extent the versions across the different modules are kept in sync.

Something like:-

Patch is clear, it should be able to be run without any worries.
Minor. I'm guessing this might have changes to fields for example but still support previous Minor versions. So added fields, not removed, for example.
Major. We can break your site, you will need to check that the fields you were using are still there.

Also how does this work across the modules. Major versions increment together (even if a specific module hasn't had a breaking change)? Or rely on compatibility requirement statements? The latter seems more correct, and more confusing at the same time.

Branches I'm agnostic. I'm quite happy to leave and have a production branch. Or change all the branch name to be correct for the versions.

Release agreement. Product group might want input into changes for a Minor. But to tag them Developers makes sense. Big thing will be making sure that the _important_ changes are well documented (apart from the changelog type changes) so it's easy for anyone updating to know what to look out for.

**Comment by ekes** [comment]
To conclude this and move it forward. As it would be good to start tagging releases asap:

{major}.{minor}.{patch} as per semantic and drupal standards.

We're building off this now. So start at 1.0.0 already for the distro itself.

Add 1.0.0-alpha/beta/RC to individual modules as appropriate to their current state.

First bump on services might be my PR to add a field?

Objections / Additions?
Otherwise I propose we start in the next day.

**Comment by stephen-cox** [comment]
Largely happy with this. Not sure about starting at 1.0.0 - this implies a certain stability which I don't think we have yet. That said, Drupal suggests the naming convention 1.0.0-alpha1 (which I find a little odd), I prefer 1.x-alpha1.

I suggest we create an 1.x-alpha1 or 1.0.0-alpha1 tag for all modules that are currently included in the profile and create a meta issue in each issue queue to discuss what needs to be done to get the module to a full release. I foresee some disagreement on what warrants an alpha, beta and RC release, but we should be able to reach some consensus on this; Drupal guidelines on this are here: https://www.drupal.org/node/1015226#pre-release

**Comment by ekes** [comment]
Don't tagged versions have to be major.minor.patch-prereleasetag otherwise they don't pattern match semantic versioning.

Stability isn't alpha across the board as we're building in upgrade paths and not breaking the api already. It might be alpha for some of the component modules though. The distribution itself however isn't enabling them.

**Comment by finnlewis** [comment]
Totally agree with the {major}.{minor}.{patch} as per semantic and Drupal standards. 
Reading that link that @stephen-cox just posted to the pre-release guidelines, I reckon we tag at
1.0.0-alpha1 
for all modules now. 

We are in the alpha phase.of the project, which may or may not mean anything to people, but fits nicely.

But more importantly, we are starting to use the distro and the primary need is a tagged release against which to pin an installed version.

So +1 for 1.0.0-alpha1 distribution release, including all sub-modules and profiles, and we can work out the parameters of increments as a next step.

@Adnan-cds @andybroomfield @cjstevens78 @OliverHannan @danchamp @tanc any strong opinions on this?

**Comment by stephen-cox** [comment]
My preferred approach would be to tag all modules in the profile as something like 0.1.0 and disregard the prerelease tags as not very semantic, but I realise that this not very Drupal.

There _may_ be a few modules we could class as beta, but I think most are, at a minimum, missing documentation and tests.

**Comment by ekes** [comment]
> My preferred approach would be to tag all modules in the profile as something like 0.1.0 and disregard the prerelease tags as not very semantic

There I do have a personal not so rational soft-spot for 0.9.0

**Comment by finnlewis** [comment]
As a follow on from this... which may feed into this decision, once we get something agreed, it would be good to get projects up on Drupal.org. At the very least the https://drupal.org/project/localgov project

So... would that be an argument for going more Drupal (even if 'less semantic').

Are there any precedents out there... Thunder? Lightning? OpenSocial?

**Comment by Adnan-cds** [comment]
> So... would that be an argument for going more Drupal (even if 'less semantic').

I thought the **new** Drupal [versioning scheme](https://www.drupal.org/node/1015226#release-tags) is totally semantic given that it says *"or for projects compatible with the latest versions of Drupal 8 and later, the format can be {major}.{minor}.{patch}. Both formats can have a pre-release suffix, like {major}.{minor}.{patch}-(alpha|beta|rc){N}"*

> There I do have a personal not so rational soft-spot for 0.9.0

Since Ekes likes 0.9.0 and Steve prefers 0.1.0, can we go for something within this range? :D

**Comment by OliverHannan** [comment]
Not sure if you meant to tag me @finnlewis but thanks as I've now had a crash course in Drupal versioning, I was guessing before.

Despite never having named one of these before, like @stephen-cox I also feel we shouldn't go straight to 1.0.0. I'm a cautious guy. 

0.9.0 feels good

**Comment by andybroomfield** [comment]
As I'm a little old fashioned, I like the old Drupal versions system, but without the 8.x bit.
We are currently developing the 1.x branch for modules, so ours should be 1.x-dev and versions 1.0.0-alpha1.
Otherwise going from 0.1 / 0.9 to 1.0 would indicate a breaking change in composer.
(Maybe for promotion, just tag it as pre-release).

**Comment by danchamp** [comment]
We also need to consider any future new development. According to semantic versioning, major version zero is for initial development where anything may change, so won't we have to adopt 0.y.z in the future, even if all current packages are considered public API stable?

**Comment by finnlewis** [comment]
Sprint review meeting notes: 
 - setting 1.0.0-alpha1 for all projects,  will follow Drupal release guidance.
 - If we start by tagging all projects with this tag, we can then review each project with a view to re-tagging or what's needed to progress
 - @Adnan-cds noted, that some projects might be beyond alpha, so we can update these next to beta for example.
 - Theme questions - what's the plan? Iniitall 1.0.0-alpha1
 - Suggestion that any new project that is a candidate for inclusion needs a 1.0.0-alpha1 tag to be included.

**Comment by stephen-cox** [comment]
All packages installed by the profile have been tagged 1.0.0-alpha1 and all composer dependencies have been set to '^1.0@alpha'. There's a merge request on the profile #147, which is more for feedback rather than anything else as I needed to update composer.json files and tag repos to get tests to pass.

**Comment by finnlewis** [comment]
Following our sprint review where we confirmed 1.0.0-alpha1 is good enough for now @stephen-cox tagged all the things, so have a starting point!

https://github.com/localgovdrupal/localgov/releases/tag/1.0.0-alpha1

Thanks everyone for the discussion!

Now we need some clear policy on how and when new releases get tagged.

**Comment by Adnan-cds** [comment]
> Now we need some clear policy on how and when new releases get tagged.

If we do not tag every time something is merge with the master branch, those new changes won't be picked up any more by composer.  So at the least, we need to bump up the PATCH number just after a merge-to-master happens.

**Comment by finnlewis** [comment]
Now that there are a couple of changes that downstream consumers (eg. Lambeth) could do with, we need to firm up the process of tagging sub modules and incrementing the release of the distrubution. 

- Coding standards issue resolved: https://github.com/localgovdrupal/localgov_alert_banner/pull/87
- Removal of patch that is no longer needed: https://github.com/localgovdrupal/localgov_geo/pull/8 

We're going to need to tag another release soon, so what aspects of the process do we need to consider.

Do we go for a weekly release perhaps - 1.0.0-alpha2 this Friday for example? Or do it more on a as needed basis?

**Comment by andybroomfield** [comment]
We also need to agree the pattern, see the release  @Adnan-cds created as 1.0.1-alpha1.
https://github.com/localgovdrupal/localgov_alert_banner/releases/tag/1.0.1-alpha1

**Comment by finnlewis** [comment]
Oh - yes, I would expect that to be 1.0.0-apha2 @Adnan-cds @andybroomfield

**Comment by finnlewis** [comment]
For example https://www.drupal.org/project/linkit_media_library/releases 

    1.0.1 - 23 July 2020
    1.0.0 - 6 July 2020
    1.0.0-alpha6 - 25 June 2020
    1.0.0-alpha5 - 23 June 2020
    1.0.0-alpha4 - 23 June 2020
    1.0.0-alpha3 - 23 June 2020
    1.0.0-alpha2 - 22 June 2020
    1.0.0-alpha1 - 21 June 2020
    1.0.x-dev - 21 June 2020

**Comment by stephen-cox** [comment]
A few thoughts about what might trigger a new tagged release for a specific module:
- A critical bug that is breaking production sites.
- A new feature.
- A build-up of minor fixes that would be good to release.

Is there anything else we might want to add to the list?

Tags should only happen when we're sure all tests are passing, so another thought, is the testing regime we have sufficient? We could certainly do with better test coverage on a module level - these are only alpha releases so that is understandable - but do we need some integration tests to ensure everything in the profile works together as well?

**Comment by Adnan-cds** [comment]
> Is there anything else we might want to add to the list?

Please note that we are no longer using "dev-master" as the version constraint.  So if we do not tag every time the master branch of a repo is updated, Composer won't pick those changes and the tests will keep failing.  Also, the MAJOR.MINOR.PATCH schema suggests that the PATCH should be incremented every time we are patching.  Considering these, I would prefer to increase the PATCH number every time the master branch is updated with any kind of bugfix big or small.  If somebody fixes a typo or anything insignificant like that then I don't see any need to touch the PATCH number.  But otherwise we should.

**Comment by Adnan-cds** [comment]
> For example https://www.drupal.org/project/linkit_media_library/releases

While I am not entirely comfortable doing 1.0.0-alpha1, 1.0.0-alpha2 etc. as shown in the example by Finn, that seem to be the way even [Composer is doing things](https://packagist.org/packages/composer/composer).  So we can't be too wrong :)

**Comment by ekes** [comment]
Tests at 'head' or specified versions of connected branches will be useful; but if something is relying on running the the test not on the module alone, but localgov checking out versions for now will need to tag.

Generally it would be good to keep tagging as soon as a branch is merged (or set of branches are merged in some unit of work); and tests pass. Worst case you know which precise version is _broken_, roll back to previous last working precise version. Generally it means you have confirmed snapshots to be working against.

**Comment by finnlewis** [comment]
Not sure, shall we try @andybroomfield - presumably if we create the new tag and delete the old one, that will be fine, no?

**Comment by ekes** [comment]
Not sure I follow this one. You don't delete tags. Each version is a permanent fixed point of known code state.

Tagging a release of a module will depending on the way dependencies are written automatically update to the next version as people run composer update; or it will require them to update composer.json version dependencies.

**Comment by finnlewis** [comment]
So in the case of a project tagged with 
1.0.1-alpha1
and
1.0.0-alpha2
which does composer determine the 'next version' to be? 
I think Andy's question is because there is already a 1.0.1-alpha1, but we are moving towards aligning with Drupal and Composer and so that should have been 1.0.0-alpha2.
I guess we can try and see what happens.
Thoughts @ekes?

**Comment by ekes** [comment]
Align to whatever the highest is by semver. So 1.0.1-alpha2 I'd say. Anything else will probably give someone who has pulled stuff grief.

**Comment by Adnan-cds** [comment]
> which does composer determine the 'next version' to be?

As Ekes has said, it should be 1.0.1-alpha2.  One of the [Semantic versioning testers](https://semver.mwl.be/#!?package=composer%2Fcomposer&version=dev-master&minimum-stability=stable) is also placing 1.0.1 before 1.0.0-alpha11

Sorry to land you into this mess.  The tests were failing yesterday, so I had to add this 1.0.1-alpha1 tag even before we agreed on a versioning format.

FAO @finnlewis @andybroomfield

**Comment by andybroomfield** [comment]
There is a reinstall plugin for composer if there are people that need to sort this out.
https://packagist.org/packages/roygoldman/composer-reinstall

Otherwise, TBH we might be close to full release on alert banner anyway. The only blocker I can think of is having a block for each type of alert banner (eg. To allow for Big Bang announcement). There are other features I'd like to add, but they can wait to 1.1.

**Comment by andybroomfield** [comment]
Closing as drupal release plan now being followed.

**Comment by ekes** [comment]
I was going to start a new issue, but I've actually thought it makes sense to keep this together, so sorry re-opening this one.

I just want to clarify how versioning: particularly major version changes, are reflected over the modules; when we bump a major, or minor version; and what constraints we're using in composer.json. This practically in light of the change made to the field naming, and the forthcoming upgrade for Drupal core.

At the moment in the composer.json for the profile project we're restricting major version updates. So ^1.0.0 it's not going to jump to 2.0.0, but will increment minor versions 1.1.0. It might be an option to only accept patch updates ~1.0.0 so 1.0.4 yes, 1.1.0 no and explicitly increment to allow, or require, minor versions. That unless we're sure that we're always going to increment the major version when anything that the profile shouldn't pull in happens.

First example. The fieldname change. If we'd incremented the major version of all the modules effected, and then increased the version in the profile composer.json and incremented it's major version. That would have required less direct communication with end users to prevent them pulling a breaking change. Now it's all alpha and breaking changes happen, but it's not something we really want or need to do if we can use versioning to help us.

I understand this would lead to the interesting situation that modules would go from 1.0.4-beta4 to 1.1.0-beta1 or something. It also means we need to get used to only those modules that require a version update get a version update.

As mentioned by @stephen-cox https://github.com/localgovdrupal/localgov/issues/139#issuecomment-702082571 it's not 'Drupal'; or rather it's not old Drupal. The alpha/beta tags are actually a bit more confusing here. 
But new Drupal semantic module numbering doesn't increment a major version of a module (like 8 to 9) with the major version of Drupal if the module itself isn't undergoing a change. So the 1.1.* versions or whatever can be for Drupal 8 and 9. It only needs to change major version if the module drops compatibility with something previously declared (like Drupal 8) or has other major changes itself. 

So looking forward to the Drupal 9 upgrade. We're should increment any modules (the profile included) where there is a non-BC change in support (D9 only basically) by a version large enough not to get pulled automatically. Those that don't change don't need a new version, or support both and don't make a major change, just need a patch or minor increment. It's might not be as humanly ascetically pretty, but it gives us more automated control via composer, so makes computer sense. This is where, particularly in alpha/beta we might prefer to increment minor version rather than major, but would have to change to restricting automatic updates to patch only.

tl;dr We should already be doing semver updates even while modules are tagged with alpha and beta in their patch section. We should increment individual modules, and profile, as required by that component. We could move to restricting to patch upgrades in composer.json

**Comment by ekes** [comment]
So after conversations in the Tech group and meets.

Decision: When code is accepted as _used_ _in_ _production_, even when the project itself is in 'beta', and the modules in development, we will use straight SemVer so production sites can manage their updates correctly. Modules then maintain _their_ major.minor versions independently as appropriate depending on changes made to them.

So policy: No change. 
Practice: tag 1.0.0 versions of modules.

**Comment by finnlewis** [comment]
Sorry @ekes  - I think we want to keep this open and use it to track the work to get all the projects tagged up ready for new branches / tagged releases. 

It probably wasn't obvious, but I updated the original issue description with the following acceptance tests:

- [ ] Agree the release tagging policy with the technical group.
- [ ] Publish and share the policy with the wider community.
- [ ] Make necessary changes to the composer.json files.
- [ ] Test this against Drupal9 or dependent modules ?

**Comment by finnlewis** [comment]
Just discussing with @ekes 

Closing this with new issues #194 #198 #199 #200
